### PR TITLE
[search] Added hashtags.

### PR DESCRIPTION
**Pull request by ygorshenin**
Added possibility to add hashtags before categorial tokens. In this case, these tokens will be used only as categorial filters, not as parts of name. This is an internal feature and should be used internally, mostly for debugging.

### CODE REVIEW

File: indexer/search_delimiters.cpp [hunk 1]
```diff
@@ -19,45 +19,57 @@ bool Delimiters::operator()(strings::UniChar c) const
   switch (c)
   {
   case 0x2116:  // NUMERO SIGN
-//  case 0x00B0:  // DEGREE SIGN
+                //  case 0x00B0:  // DEGREE SIGN
```
Comment (mpimenov) [inline indexer/search_delimiters.cpp]: Было лучше.

File: search/search_query.cpp [hunk 2]
```diff
@@ -461,9 +469,42 @@ void Query::SetQuery(string const & query)
   ASSERT(m_tokens.empty(), ());
   ASSERT(m_prefix.empty(), ());
 
-  // Split input query by tokens with possible last prefix.
+  // Following code splits input query by delimiters except hash tags
```
Comment (mpimenov) [inline search/search_query.cpp]: s/hash tags/hashtags/

File: search/search_query.cpp [hunk 3]
```diff
@@ -461,9 +469,42 @@ void Query::SetQuery(string const & query)
...
+  // first, and then splits result tokens by hash tags. The reason is
+  // to retrieve all tokens that starts with a single hash tag sign
```
Comment (mpimenov) [inline search/search_query.cpp]: s/starts/start/

File: search/search_query.cpp [hunk 4]
```diff
@@ -461,9 +469,42 @@ void Query::SetQuery(string const & query)
   ASSERT(m_tokens.empty(), ());
   ASSERT(m_prefix.empty(), ());
 
-  // Split input query by tokens with possible last prefix.
+  // Following code splits input query by delimiters except hash tags
+  // first, and then splits result tokens by hash tags. The reason is
```
Comment (mpimenov) [inline search/search_query.cpp]: s/reason/goal/

File: search/search_query.cpp [hunk 5]
```diff
@@ -206,6 +206,15 @@ void UpdateNameScore(vector<strings::UniString> const & tokens, TSlice const & s
     bestCoverage = coverage;
   }
 }
+
+inline bool IsHashed(strings::UniString const & s) { return !s.empty() && s[0] == '#'; }
```
Comment (mpimenov) [inline search/search_query.cpp]: s/IsHashed/IsHashtagged/ !

File: search/search_query.cpp [hunk 6]
```diff
@@ -206,6 +206,15 @@ void UpdateNameScore(vector<strings::UniString> const & tokens, TSlice const & s
...
+
+inline strings::UniString RemoveHash(strings::UniString const & s)
```
Comment (mpimenov) [inline search/search_query.cpp]: s/RemoveHash/RemoveHashtag/

File: indexer/search_delimiters.cpp (hunk 1 above)
Comment (ygorshenin) [inline indexer/search_delimiters.cpp]: Done

File: search/search_query.cpp (hunk 2 above)
Comment (ygorshenin) [inline search/search_query.cpp]: Done

File: search/search_query.cpp (hunk 3 above)
Comment (ygorshenin) [inline search/search_query.cpp]: Done

File: search/search_query.cpp (hunk 4 above)
Comment (ygorshenin) [inline search/search_query.cpp]: Done

File: search/search_query.cpp (hunk 5 above)
Comment (ygorshenin) [inline search/search_query.cpp]: Done

File: search/search_query.cpp (hunk 6 above)
Comment (ygorshenin) [inline search/search_query.cpp]: Done

**Comment by ygorshenin** [comment]
PTAL

File: search/search_query.cpp (hunk 2 above)
Comment (mpimenov) [inline search/search_query.cpp]: ping on this comment

File: search/search_query.cpp [hunk 7]
```diff
@@ -461,9 +469,42 @@ void Query::SetQuery(string const & query)
...
-  SplitUniString(NormalizeAndSimplifyString(query), MakeBackInsertFunctor(m_tokens), delims);
+  {
+    buffer_vector<strings::UniString, 32> subTokens;
+    for (auto const & token : tokens)
+    {
+      size_t numHashes = 0;
+      for (; numHashes < token.size() && token[numHashes] == '#'; ++numHashes)
+        ;
+
+      // Splits |tokens[i]| by hash signs, because all other
```
Comment (mpimenov) [inline search/search_query.cpp]: s/tokens[i]/token/

File: search/search_integration_tests/search_query_v2_test.cpp [hunk 8]
```diff
@@ -467,5 +458,31 @@ UNIT_CLASS_TEST(SearchQueryV2Test, TestPostcodes)
...
+  auto wonderlandId = BuildCountry(countryName, [&](TestMwmBuilder & builder)
+                                   {
+                                     builder.Add(atm);
+                                   });
+
+  SetViewport(m2::RectD(m2::PointD(-0.5, -0.5), m2::PointD(0.5, 0.5)));
+  {
+    TRules rules = {ExactMatch(wonderlandId, atm)};
+    TEST(ResultsMatch("atm", rules), ());
+    TEST(ResultsMatch("#atm", rules), ());
```
Comment (mpimenov) [inline search/search_integration_tests/search_query_v2_test.cpp]: Мб добавить больше тестов?
Например, у фичи есть в названии "atm", но тип не "atm".
И наоборот.

Comment (ygorshenin) [inline search/search_integration_tests/search_query_v2_test.cpp]: Done

File: search/search_query.cpp (hunk 7 above)
Comment (ygorshenin) [inline search/search_query.cpp]: Done

File: search/search_integration_tests/search_query_v2_test.cpp [hunk 9]
```diff
@@ -467,5 +458,54 @@ UNIT_CLASS_TEST(SearchQueryV2Test, TestPostcodes)
...
+
+    TestSearchRequest request(m_engine, params, m_viewport);
+    request.Wait();
+
+    TEST(MatchResults(m_engine, rules, request.Results()), ());
+    for (auto const & result : request.Results())
+    {
+      auto const & info = result.GetRankingInfo();
+
+      // Token with a hashtag should not participate in a name-score
```
Comment (mpimenov) [inline search/search_integration_tests/search_query_v2_test.cpp]: s/a name-score/name-score/

**Comment by mpimenov** [comment]
LGTM after fix

File: search/search_integration_tests/search_query_v2_test.cpp (hunk 9 above)
Comment (ygorshenin) [inline search/search_integration_tests/search_query_v2_test.cpp]: Done

**Pull request merged**